File: sdk/ios/Sources/IAPStackApple/IAPStackRetryPolicy.swift
import Foundation

/// Retry settings for idempotent IAPStack operations.
public struct IAPStackRetryPolicy: Sendable {
  /// Creates a bounded full-jitter exponential backoff policy.
  public init(
    maxAttempts: Int = 3,
    baseDelay: TimeInterval = 0.25,
    maxDelay: TimeInterval = 2.0,
  ) {
    self.maxAttempts = maxAttempts
    self.baseDelay = baseDelay
    self.maxDelay = maxDelay
  }

  /// Total attempts, including the initial request.
  public let maxAttempts: Int

  /// Upper delay bound before the second attempt.
  public let baseDelay: TimeInterval

  /// Upper delay bound.
  public let maxDelay: TimeInterval

  /// Returns a full-jitter delay after a failed one-based [attempt].
  public func delayAfter(attempt: Int, randomValue: Double) -> TimeInterval {
    precondition(attempt > 0, "attempt must be positive")
    precondition(randomValue >= 0.0 && randomValue <= 1.0, "randomValue must be in [0, 1]")
    guard baseDelay > 0 else { return 0 }
    let exponent = min(max(attempt - 1, 0), 20)
    let multiplier = 1 << exponent
    let capped = min(baseDelay * Double(multiplier), maxDelay)
    return capped * randomValue
  }

  /// Throws `IAPStackSDKError.configurationError` when bounds are unsafe.
  public func validate() throws {
    if maxAttempts < 1 || maxAttempts > 5 {
      throw IAPStackSDKError.configurationError(
        message: "maxAttempts must be between 1 and 5",
      )
    }
    if baseDelay < 0 || maxDelay < 0 || baseDelay > maxDelay {
      throw IAPStackSDKError.configurationError(
        message: "retry delays must be non-negative and ordered",
      )
    }
  }
}

File: sdk/android/core/src/main/kotlin/com/iapstack/core/IapStackRetryPolicy.kt
package com.iapstack.core

import kotlin.time.Duration
import kotlin.time.Duration.Companion.microseconds
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds

/**
 * Full-jitter exponential backoff policy for idempotent read-write operations.
 */
data class IapStackRetryPolicy(
  val maxAttempts: Int = 3,
  val baseDelay: Duration = 250.milliseconds,
  val maxDelay: Duration = 2.seconds,
) {
  fun validate() {
    if (maxAttempts < 1 || maxAttempts > 5) {
      throw IllegalArgumentException("maxAttempts must be between 1 and 5")
    }
    if (baseDelay.isNegative() || maxDelay.isNegative() || baseDelay > maxDelay) {
      throw IllegalArgumentException("retry delays must be non-negative and ordered")
    }
  }

  /**
   * Returns a bounded full-jitter delay after a one-based failed attempt.
   */
  fun delayAfter(
    attempt: Int,
    randomValue: Double,
  ): Duration {
    if (randomValue !in 0.0..1.0) {
      throw IllegalArgumentException("randomValue must be between 0 and 1")
    }
    if (baseDelay == Duration.ZERO) {
      return Duration.ZERO
    }
    val exponent = (attempt - 1).coerceAtLeast(0).coerceAtMost(20)
    val multiplier = 1L shl exponent
    val capMicros =
      (baseDelay.inWholeMicroseconds * multiplier).coerceAtMost(maxDelay.inWholeMicroseconds)
    return (capMicros * randomValue).toLong().microseconds
  }
}
